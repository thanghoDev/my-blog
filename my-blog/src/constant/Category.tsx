const CATEGORY = 'category';
const DETAIL = 'detail';
export { CATEGORY, DETAIL };
