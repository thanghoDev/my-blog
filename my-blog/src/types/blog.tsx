export type Blog = {
  id: number;
  title: string;
  description: string;
  category: string;
  view: number;
  alt: string;
  images: string;
};
