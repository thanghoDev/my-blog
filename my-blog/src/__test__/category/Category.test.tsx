import React from 'react';
import { useRouter } from 'next/router';
import { render, screen } from '@testing-library/react';
import Category, {
  getStaticPaths,
  getStaticProps,
} from '@/pages/category/[category]/[page]';
import { Blog } from '@/types/blog';

jest.mock('next/router', () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

describe('testing Category page', () => {
  test('testing getStaticPaths', async () => {
    const data = [
      {
        id: '1',
        category: 'business',
      },
    ];
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => data,
      })
    );
    const response = await (getStaticPaths as jest.Mock)();
    const paths = [{ params: { category: 'business', page: '1' } }];
    expect(response).toEqual({ paths, fallback: true });
    expect(fetch).toHaveBeenCalled();
  });

  test('testing getStaticProps', async () => {
    const categories = [
      {
        id: 1,
        title: 'title',
        description: 'description',
        category: 'business',
        view: 111,
        alt: 'alt',
        images: 'https://i.ibb.co/10hj3h1/carousel.jpg',
        content: 'content',
      },
    ];
    const category = 'business';
    const page = '1';
    const trending = [
      {
        id: 1,
        title: 'title',
        description: 'description',
        category: 'business',
        view: 111,
        alt: 'alt',
        images: 'https://i.ibb.co/10hj3h1/carousel.jpg',
        content: 'content',
      },
    ];
    const allCategory = [
      {
        id: 1,
        title: 'title',
        description: 'description',
        category: 'business',
        view: 111,
        alt: 'alt',
        images: 'https://i.ibb.co/10hj3h1/carousel.jpg',
        content: 'content',
      },
    ];
    const totalPage = [1];

    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => categories,
      })
    );
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => trending,
      })
    );
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => allCategory,
      })
    );

    const response = await getStaticProps({
      params: { category: 'business', page: '1' },
    });
    expect(response).toEqual({
      props: {
        categories: categories,
        totalPage: totalPage,
        trending: trending,
        category: category,
        page: page,
      },
      revalidate: 300,
    });
    expect(fetch).toHaveBeenCalled();
  });
  test('testing getStaticProps render empty', async () => {
    const categories: Blog[] = [];
    const trending: Blog[] = [];
    const allCategory: Blog[] = [];
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => categories,
      })
    );
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => trending,
      })
    );
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => allCategory,
      })
    );
    const response = await getStaticProps({
      params: { category: 'searchValue', page: '1' },
    });
    expect(response).toEqual({
      notFound: true,
    });
    expect(fetch).toHaveBeenCalled();
  });

  test('render category page', () => {
    const mockRouter = {
      isFallback: false,
    };
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
    const category = 'category';
    const page = '1';
    const totalPage = [1, 2];
    const trending = [
      {
        id: 1,
        title: 'title',
        description: 'description',
        category: 'health',
        view: 150,
        images: 'https://i.ibb.co/10hj3h1/health.jpg',
        alt: 'health',
        content: 'content',
      },
    ];
    const categories = [
      {
        id: 1,
        title: 'title',
        description: 'description',
        category: 'PostList',
        view: 150,
        images: 'https://i.ibb.co/10hj3h1/PostList.jpg',
        alt: 'PostList',
        content: 'content',
      },
    ];
    const { container } = render(
      <Category
        category={category}
        categories={categories}
        trending={trending}
        page={page}
        totalPage={totalPage}
      />
    );
    expect(container).toMatchSnapshot();
    const title = screen.getByRole('heading', {
      name: /category/i,
    });
    expect(title).toBeInTheDocument();
  });
  test('render loading spinner', () => {
    const mockRouter = {
      isFallback: true,
    };
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
    render(
      <Category
        category={''}
        page={''}
        totalPage={[]}
        categories={[]}
        trending={[]}
      />
    );
    const spinner = screen.getByTestId('spinner');
    expect(spinner).toBeInTheDocument();
  });
});
