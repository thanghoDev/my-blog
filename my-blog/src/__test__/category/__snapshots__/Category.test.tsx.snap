// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Category page render category page 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="d-flex"
    >
      <div
        class="flex-column row"
      >
        <span>
          Categories
        </span>
        <h2
          class="text-capitalize text-decoration-underline"
        >
          category
        </h2>
        <div
          class="mt-5 col"
        >
          <div
            class="d-flex mb-3"
          >
            <div
              class="me-3 w-100"
            >
              <a
                class="fs-5 text-black text-capitalize postTitle text-decoration-none"
                href="/detail/1"
              >
                title
              </a>
              <p
                class="description"
              >
                description
              </p>
              <p
                class="text-capitalize category"
              >
                Dave Rogers 
                <code
                  class="text-secondary"
                >
                  in
                </code>
                 
                PostList
              </p>
            </div>
            <span
              style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
            >
              <span
                style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
              >
                <img
                  alt=""
                  aria-hidden="true"
                  src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%27248%27%20height=%27155%27/%3e"
                  style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
                />
              </span>
              <img
                alt="PostList"
                class="w-25"
                data-nimg="intrinsic"
                decoding="async"
                src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
              />
              <noscript />
            </span>
          </div>
        </div>
      </div>
      <div
        class="mt-5 ps-4 col-md-4"
      >
        <h2
          class="mb-5 text-decoration-underline text-capitalize fs-5"
        >
          trending
        </h2>
        <div
          class="mt-3 d-flex"
        >
          <h3>
            0
            1
          </h3>
          <div
            class="ms-4"
          >
            <a
              class="text-black text-capitalize text-decoration-none undefined"
              href="/detail/1"
            >
              title
            </a>
            <p
              class="text-capitalize category mt-3"
            >
              Dave Rogers 
              <code
                class="text-secondary"
              >
                in
              </code>
               
              <a
                class="text-black text-decoration-none"
                href="/category/health"
              >
                health
              </a>
            </p>
          </div>
        </div>
        <a
          class="text-uppercase link-success text-decoration-none"
          href="/#"
        >
          See all trends
          <i
            class="bi bi-chevron-right"
          />
        </a>
      </div>
    </div>
    <ul
      class="pagination"
    >
      <li
        class="disabled page-item"
      >
        <a
          class="page-link"
          href="/category/category/1"
        >
          1
        </a>
      </li>
      <li
        class="page-item"
      >
        <a
          class="page-link"
          href="/category/category/2"
        >
          2
        </a>
      </li>
    </ul>
  </div>
  <div
    class="bg-light pt-5 pb-5 mt-5"
  >
    <div
      class="container pb-5 pt-5"
    >
      <form
        action="#"
        class="row align-items-center"
      >
        <div
          class="col-md-5 mr-auto"
        >
          <h2
            class="text-capitalize"
          >
            newsletter subscribe
          </h2>
          <p
            class="description"
          >
            Lorem ipsum dolor sit amet, consectetur adipisicing elit. Perferendis aspernatur ut at quae omnis pariatur obcaecati possimus nisi ea iste!
          </p>
        </div>
        <div
          class="col-md-6 ml-auto d-flex gap-2"
        >
          <input
            class="form-control rounded-pill"
            placeholder="Enter your email"
            type="email"
          />
          <button
            class="btn btn-outline-success border border-light rounded-circle bg-dark"
            type="button"
          >
            <i
              class="bi bi-send text-light"
            />
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
